// appdynamics/lib/probes/ioredis-probe.js
/*
Copyright (c) AppDynamics, Inc., and its affiliates
2015
All Rights Reserved
 */
'use strict';


function IoredisProbe(agent) {
  this.agent = agent;

  this.packages = ['ioredis'];
}
exports.IoredisProbe = IoredisProbe;



IoredisProbe.prototype.attach = function(obj) {
  var self = this;

  if(obj.__appdynamicsProbeAttached__) return;
  obj.__appdynamicsProbeAttached__ = true;
  self.agent.on('destroy', function() {
    if(obj.__appdynamicsProbeAttached__) {
      delete obj.__appdynamicsProbeAttached__;
      proxy.release(obj.Cluster.prototype.sendCommand);
      proxy.release(obj.prototype.sendCommand);
    }
  });

  var proxy = self.agent.proxy;
  var profiler = self.agent.profiler;
  var clusters = {};


  proxy.before(obj.Cluster.prototype, "sendCommand", function(obj) {
    var serverPool = [];
    if(Array.isArray(obj.startupNodes)) {
      obj.startupNodes.forEach(function(node) {
        var address = node.host + ':' + node.port;
        serverPool.push(address);
        clusters[address] = serverPool;
      });
    }
  });


  proxy.around(obj.prototype, "sendCommand", function(obj, args, locals) {
    var redis = obj;
    var command = args[0];
    var commandName = command.name;
    var commandArgs = command.args;
    var address = redis.options.host + ':' + redis.options.port;
    locals.time = profiler.time();

    var serverPool = clusters[address];
    if(serverPool) {
      address = serverPool.join('\n');
    }

    var supportedProperties = {
      'SERVER POOL': address,
      'VENDOR': 'REDIS'
    };

    locals.exitCall = profiler.createExitCall(locals.time, {
      exitType: 'EXIT_CACHE',
      backendName: 'Redis',
      supportedProperties: supportedProperties,
      command: commandName,
      commandArgs: profiler.sanitize(commandArgs),
      stackTrace: profiler.stackTrace()
    });

    if (!locals.exitCall) return;

    if(command.callback && typeof(command.callback) === 'function') {
      locals.methodHasCb = true;
      proxy.before(command, 'callback', function(obj, args) {
        complete(args, locals);
      });

      proxy.before(command.promise, '_promise0', function(obj, args) {
        complete(args, locals);
      });
    }
  }, after);

  function after(obj,args, ret, locals) {
    if (locals.methodHasCb)
      return;
    if (!ret || !ret.__appdynamicsIsPromiseResult__)
      complete(null, locals);
    else if (ret.error)
      complete(ret.error, locals);
    else
      complete(null, locals);
  }

  function complete(err, locals) {
    if (!locals.exitCall) return;
    if (!locals.time.done()) return;

    var error = self.agent.proxy.getErrorObject(err);
    profiler.addExitCall(locals.time, locals.exitCall, error);
  }
};
