// appdynamics/lib/probes/mysql-probe.js
/*
Copyright (c) AppDynamics, Inc., and its affiliates
2015
All Rights Reserved
 */
'use strict';


function MysqlProbe(agent) {
  this.agent = agent;

  this.packages = ['mysql'];
}
exports.MysqlProbe = MysqlProbe;

function prepareExitCall(config, client, proxy, profiler) {
  var supportedProperties = {
    'HOST': config.host,
    'PORT': config.port,
    'DATABASE': config.database,
    'VENDOR': 'MYSQL'
  };
  if (client.query.__appdynamicsProbeAttached__) return;

  proxy.around(client, 'query', function(obj, args, locals) {
    var command, params;
    var trace = profiler.stackTrace();
    locals.time = profiler.time();

    function processQueryCb(obj, args) {
      complete(args, locals);
    }

    function createExitCall() {
      locals.exitCall = profiler.createExitCall(locals.time, {
        exitType: 'EXIT_DB',
        backendName: 'MySQL',
        supportedProperties: supportedProperties,
        command: profiler.sanitize(command),
        commandArgs: profiler.sanitize(params),
        user: config.user,
        stackTrace: trace,
        isSql: true
      });
    }

    if (args[0] && args[0].__appdCbFnHolder) {
      command = args[0].sql;
      params = args[0].values;
      locals.methodHasCb = true;
      createExitCall();
      proxy.before(args[0].__appdCbFnHolder, 'triggerOnCb', processQueryCb);
    } else {
      command = args.length > 0 ? args[0] : undefined;
      params = args.length > 1 && Array.isArray(args[1]) ? args[1] : undefined;
      createExitCall();
      locals.methodHasCb = proxy.callback(args, -1, processQueryCb);
    }
  }, after);

  client.query.__appdynamicsProbeAttached__ = true;

  function after(obj, args, ret, locals) {
    if (locals.methodHasCb) return;
    if (!ret || !ret.__appdynamicsIsPromiseResult__)
      complete(null, locals);
    else if (ret.error)
      complete(ret.error, locals);
    else {
      complete(null, locals);
    }
  }

  function complete(err, locals) {
    if (!locals.exitCall) return;
    if (!locals.time.done()) return;

    var error = proxy.getErrorObject(err);
    profiler.addExitCall(locals.time, locals.exitCall, error);
  }
}

MysqlProbe.prototype.attach = function(obj) {
  var self = this;
  var proxy = self.agent.proxy;
  var profiler = self.agent.profiler;
  var cmds = ['createClient', 'createConnection', 'createPool'];
  var mysqlConnection;

  if(obj.__appdynamicsProbeAttached__) return;
  obj.__appdynamicsProbeAttached__ = true;
  self.agent.on('destroy', function() {
    if(obj.__appdynamicsProbeAttached__) {
      delete obj.__appdynamicsProbeAttached__;
      cmds.forEach(function(createCmd) {
        if (obj[createCmd])
          proxy.release(obj[createCmd]);
      });
      if (mysqlConnection)
        delete mysqlConnection.query.__appdynamicsProbeAttached__;
    }
  });

  cmds.forEach(function(createCmd) {
    proxy.after(obj, createCmd, function(obj, args, ret) {
      if (createCmd === 'createPool') {
        proxy.around(ret, 'getConnection', function(obj, args, locals) {
          locals.methodHasCb = proxy.callback(args, -1, function(obj, args) {
            if (args[0]) return;
            if (args[1]) {
              mysqlConnection = args[1];
              var config = mysqlConnection.config;
              prepareExitCall(config, mysqlConnection, proxy, profiler);
            }
          });
        }, function(obj, args, ret, locals) {
          if (locals.methodHasCb) return;
          if (!ret || !ret.__appdynamicsIsPromiseResult__) return;
          if (ret.error) return;
          else {
            mysqlConnection = ret.data;
            var config = mysqlConnection.config;
            prepareExitCall(config, mysqlConnection, proxy, profiler);
          }
        });

        proxy.around(ret, 'query', function(obj, args, locals) {
          locals.triggerOnCb = function() {
            self.agent.logger.info('Placeholder function');
          };
          if (typeof args[0] === 'function') {
            proxy.before(args, '0', function() {
              locals.triggerOnCb();
            });
          } else if (typeof args[1] === 'function') {
            proxy.before(args, '1', function() {
              locals.triggerOnCb();
            });
          } else {
            proxy.before(args, '2', function() {
              locals.triggerOnCb();
            });
          }
        }, function(obj, args, query, locals) {
          query.__appdCbFnHolder = locals;
        });
      } else {
        mysqlConnection = ret;
        var config = (createCmd === 'createClient' ? mysqlConnection : mysqlConnection.config);
        if (!config) {
          return;
        }
        prepareExitCall(config, mysqlConnection, proxy, profiler);
      }
    });
  });
};
